### PR TITLE
List artifact provides and depends from the CLI. 

### PR DESCRIPTION
**Pull request by oleorhagen**
Changelog: Title

Signed-off-by: Ole Petter <ole.orhagen@cfengine.com>

### CODE REVIEW

File: cli/mender-artifact/read.go [hunk 1]
```diff
@@ -105,3 +110,59 @@ func readArtifact(c *cli.Context) error {
...
 }
+
+func listDepends(c *cli.Context) error {
+	if c.NArg() == 0 {
+		return cli.NewExitError("Nothing specified, nothing read. \nMaybe you wanted"+
+			" to say 'artifacts read-depends <pathspec>'?", errArtifactInvalidParameters)
+	}
+
+	f, err := os.Open(c.Args().First())
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (kacf) [reply cli/mender-artifact/read.go]: I know what you mean, I've had a few similar cases in my update module branch.

File: artifact/metadata.go [hunk 2]
```diff
@@ -284,12 +285,46 @@ type ArtifactDepends struct {
 	UpdateTypesSupported []string `json:"update_types_supported,omitempty"`
 }
 
+func (a *ArtifactDepends) String() string {
+	funcMap := template.FuncMap{
+		"join": strings.Join,
+	}
+	tmpl := `{{if .ArtifactName}}{{"\tArtifact Name:\n"}}{{range .ArtifactName}}{{println "\t\t-" .}}{{end}}{{end}}
+{{- if .CompatibleDevices}}{{"\tCompatible Devices:\n"}}{{range .CompatibleDevices}}{{println "\t\t-" .}}{{end}}{{end}}
```
[the remark this answers is not in the document]
Comment (kacf) [reply artifact/metadata.go]: I think the template interface is nice, but I think the function should rather be named `PrettyPrint()` or `FormattedString()` or something. `String()` has a special meaning in Go, which you have taken advantage of in the `Printf` below, but this function is doing more than just converting the data to a string, it's adding a lot of formatting data (like multiple lines) which is only relevant in the context of the command line. If used somewhere else, I think this `String()` implementation would give unexpected results, for example in logging.

File: cli/mender-artifact/read.go [hunk 3]
```diff
@@ -83,7 +83,12 @@ func readArtifact(c *cli.Context) error {
 	fmt.Printf("  Format: %s\n", info.Format)
 	fmt.Printf("  Version: %d\n", info.Version)
 	fmt.Printf("  Signature: %s\n", sigInfo)
-	fmt.Printf("  Compatible devices: '%s'\n", r.GetCompatibleDevices())
+	if info.Version == 3 {
+		fmt.Printf("\n%s", r.GetArtifactDepends())
```
[the remark this answers is not in the document]
Comment (kacf) [reply cli/mender-artifact/read.go]: The output is good, just a small concern about the calling of the string formatter, as per my other comment.

File: cli/mender-artifact/main.go [hunk 4]
```diff
@@ -252,9 +252,27 @@ func run() error {
...
+		Description: "This command lists all the provides of a Mender image or an artifact",
+	}
+
+	readDepends := cli.Command{
+		Name:        "read-depends",
+		Usage:       "List all artifact-depends",
+		Action:      listDepends,
+		UsageText:   "mender-artifact read-depends <pathspec>",
+		Description: "This command lists all the depends of a Mender image or an artifact",
+	}
```
[the remark this answers is not in the document]
Comment (oleorhagen) [reply cli/mender-artifact/main.go]: I'd vote for splitting them up. However, is this jumping the gun, and can be taken care of at a later point, when we know more what the final version is gonna look  like?

[the remark this answers is not in the document]
Comment (kacf) [reply cli/mender-artifact/main.go]: Yeah, I'm ok with that. I will probably make some modifications once I rebase anyway.

**Pull request closed without merging**